### PR TITLE
Refactor SPDZ to use the crypto store

### PR DESCRIPTION
**Pull request by LaRiffle**
## Description
This PR aligns SPDZ with the FSS protocol to use a unified way of generating crypto primitives for those protocol. This will also benefit SecureNN.

**🛑     WARNING: This PR removes support for SMPC with > 2 workers**

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: syft/frameworks/torch/mpc/beaver.py [hunk 1]
```diff
@@ -1,45 +1,60 @@
...
 
-from syft.workers.abstract import AbstractWorker
 
-
-def request_triple(
-    crypto_provider: AbstractWorker,
-    cmd: Callable,
+def build_triple(
+    op: str,
+    shape: Tuple[th.Size],
```
Comment (youben11) [inline syft/frameworks/torch/mpc/beaver.py:7]: I think you mention that shape is a tuple of two shapes
```suggestion
    shape: Tuple[th.Size, th.Size],
```

Comment (Yugandhartripathi) [reply syft/frameworks/torch/mpc/beaver.py:7]: Yes, it should be tuple of two shapes.

File: syft/frameworks/torch/mpc/primitives.py [hunk 2]
```diff
@@ -191,3 +237,34 @@ def build_separate_fss_keys(n_party: int, n_instances: int = 100):
...
+        """
+
+        def build_separate_triples(n_party: int, n_instances: int, **kwargs):
+            assert n_party == 2, (
+                "Only 2 workers supported for the moment. "
+                "Please fill an issue if you have an urgent need."
+            )
+            shapes = kwargs["shapes"]
+            if not isinstance(shapes, list):
+                assert len(shapes) == 2
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/primitives.py:253]: Q: Shouldn't this assert be after the ```if```?

File: syft/frameworks/torch/mpc/spdz.py [hunk 3]
```diff
@@ -1,57 +1,164 @@
...
+    cmd = th.matmul
+    delta_b = cmd(delta, b)
+    a_epsilon = cmd(a, epsilon)
+    delta_epsilon = cmd(delta, epsilon)
+    return core_id, delta_b, a_epsilon, delta_epsilon
+
+
+# share level
+@allow_command
+def spdz_compute(j, delta, epsilon, op, dtype, torch_dtype, field):
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/spdz.py:51]: Q: Could you add a description? Not sure what is ```j```

File: syft/frameworks/torch/mpc/spdz.py [hunk 4]
```diff
@@ -1,57 +1,164 @@
...
-        j = sy.MultiPointerTensor(children=[j1, j0])
+    if not asynchronous:
+        shares = []
+        for i, location in enumerate(locations):
+            args = (th.LongTensor([i]), delta, epsilon, op, dtype, torch_dtype, field)
+            share = remote(spdz_compute, location=location)(*args, return_value=False)
+            shares.append(share)
     else:
-        j = sy.MultiPointerTensor(children=[j1] + list(j0.child.values()))
+        print("async spdz")
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/spdz.py:149]: TODO: Remove print

File: syft/workers/base.py [hunk 5]
```diff
@@ -481,6 +482,16 @@ def garbage(self, object_id, location):
             self.send_msg(ForceObjectDeleteMessage(trash[location.id][1]), location)
             trash[location.id] = (time.time(), [])
 
+    async def async_dispatch(self, workers, commands, return_value=False):
```
Comment (gmuraru) [inline syft/workers/base.py:485]: GREAT!

File: syft/frameworks/torch/mpc/primitives.py [hunk 6]
```diff
@@ -26,11 +28,15 @@ def __init__(self, owner: AbstractWorker):
         """
         self.fss_eq: list = []
         self.fss_comp: list = []
+        self.mul: list = defaultdict(list)
```
[the remark this answers is not in the document]
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/primitives.py:31]: Correct! dict I guess

File: syft/frameworks/torch/mpc/spdz.py (hunk 3 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/spdz.py:51]: Sure!

File: syft/frameworks/torch/mpc/primitives.py (hunk 2 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/primitives.py:253]: not really: shapes should actually be a list of pair of shapes. Will make it clear in the code

**Pull request merged**